datajoint/external.py

```diff
@@ -45,44 +45,82 @@ def __init__(self, connection, store=None, database=None):
+        if self.spec['protocol'] == 's3':
+            self.s3.put('/'.join((self.database, '/'.join(subfold(uuid.hex, self.spec['subfolding'])), uuid.hex)), blob)
+        else:
+            remote_file = os.path.join(os.path.join(
+                self.spec['location'], self.database, *subfold(uuid.hex, self.spec['subfolding'])), uuid.hex)
+            safe_write(remote_file, blob)
         # insert tracking info
         self.connection.query(
-            "INSERT INTO {tab} (hash, size) VALUES (%s, {size}) "
-            "ON DUPLICATE KEY UPDATE timestamp=CURRENT_TIMESTAMP".format(
-                tab=self.full_table_name, size=len(blob)), args=(blob_hash.bytes,))
-        return blob_hash
+            "INSERT INTO {tab} (hash, size) VALUES (%s, {size}) ON DUPLICATE KEY "
+            "UPDATE timestamp=CURRENT_TIMESTAMP".format(
+                tab=self.full_table_name, size=len(blob)), args=(uuid.bytes,))
+        return uuid
+
+    def fput(self, local_filepath):
+        """
+        Raise exception if an external entry already exists with a different contents checksum.
+        Otherwise, copy (with overwrite) file to remote and
+        If an external entry exists with the same checksum, then no copying should occur
+        """
+        local_folder = os.path.dirname(local_filepath)
+        stage_folder = os.path.join(os.path.abspath(self.spec['stage']), '')
+        if not local_folder.startswith(stage_folder):
+            raise DataJointError('The path {path} is not in stage {stage}'.format(
+                path=local_folder, stage=stage_folder))
+        relative_filepath = local_filepath[len(stage_folder):]
+        uuid = uuid_from_buffer(init_string=relative_filepath)
+        contents_hash = uuid_from_file(local_filepath)
+
+        # check if the remote file already exists and verify that it matches
+        check_hash = (self & {'hash': uuid}).fetch('contents_hash')
+        if check_hash:
+            # the tracking entry exists, check that it's the same file as before
+            if contents_hash != check_hash[0]:
+                raise DataJointError(
+                    "A different version of '{file}' has already been placed.".format(file=relative_filepath))
+        else:
+            # upload the file and create its tracking entry
+            if self.spec['protocol'] == 's3':
+                self.s3.fput(relative_filepath, local_filepath, contents_hash=str(contents_hash))
+            else:
+                remote_file = os.path.join(self.spec['location'], relative_filepath)
+                safe_copy(local_filepath, remote_file, overwrite=True)
+            self.connection.query(
+                "INSERT INTO {tab} (hash, size, filepath, contents_hash) VALUES (%s, {size}, '{filepath}', %s)".format(
+                    tab=self.full_table_name, size=os.path.getsize(local_filepath),
+                    filepath=relative_filepath), args=(uuid.bytes, contents_hash.bytes))
+        return uuid
```

Comment:
Don't see an implementation that signals on-going changes to a file (similar to vim swap file). This would prevent collisions if multiple users are updating the same entry.

Reply:
Per discussion, need to research S3/Minio behavior on simultaneous updates and develop a test for this. Currently, a transaction is started prior to the copy, however, it is not committed so there is a risk.